cpp/daal/src/algorithms/kmeans/oneapi/kmeans_dense_lloyd_batch_kernel_ucapi.h

```diff
@@ -47,59 +47,56 @@ class KMeansDenseLloydBatchKernelUCAPI : public Kernel
+                                  services::Status * st);
+
+    void mergePartialCandidates(uint32_t nClusters, services::Status * st);
+
+    void partialReduceCentroids(const services::Buffer<algorithmFPType> & data, const services::Buffer<int> & assignments, uint32_t blockSize,
+                                uint32_t nClusters, uint32_t nFeatures, uint32_t doReset, services::Status * st);
+
+    void mergeReduceCentroids(const services::Buffer<algorithmFPType> & centroids, uint32_t nClusters, uint32_t nFeatures, services::Status * st);
+
+    void updateObjectiveFunction(const services::Buffer<int> & assignments, const services::Buffer<algorithmFPType> & objFunction, uint32_t blockSize,
                                  uint32_t nClusters, uint32_t doReset, services::Status * st);
+    void getNumEmptyClusters(uint32_t nClusters, services::Status * st);
+    void buildProgram(oneapi::internal::ClKernelFactoryIface & kernelFactory, uint32_t nClusters, daal::services::Status * st);
+    services::Status setEmptyClusters(NumericTable * const ntData, uint32_t nRows, uint32_t nClusters, uint32_t nFeatures,
+                                      services::Buffer<algorithmFPType> & outCentroids, algorithmFPType & objFuncCorrection);
+    services::Status initializeBuffers(uint32_t nClusters, uint32_t nFeatures, uint32_t blockSize);
+    services::Status getBlockSize(uint32_t nRows, uint32_t nClusters, uint32_t nFeatures, uint32_t & blockSize);
     uint32_t getCandidatePartNum(uint32_t nClusters);
     uint32_t getWorkgroupsCount(uint32_t rows);
     uint32_t getComputeSquaresWorkgroupsCount(uint32_t nFeatures);
     const char * getComputeSquaresKernelName(uint32_t nFeatures);
-    const char * getBuildOptions(uint32_t nClusters);
+    services::String getBuildOptions(uint32_t nClusters);
+
+    oneapi::internal::UniversalBuffer _dataSq;
+    oneapi::internal::UniversalBuffer _centroidsSq;
+    oneapi::internal::UniversalBuffer _distances;
+    oneapi::internal::UniversalBuffer _mindistances;
+    oneapi::internal::UniversalBuffer _candidates;
+    oneapi::internal::UniversalBuffer _candidateDistances;
+    oneapi::internal::UniversalBuffer _partialCandidates;
+    oneapi::internal::UniversalBuffer _partialCandidateDistances;
+    oneapi::internal::UniversalBuffer _partialCentroids;
+    oneapi::internal::UniversalBuffer _partialCentroidsCounters;
+    oneapi::internal::UniversalBuffer _numEmptyClusters;
 
     const uint32_t _maxWorkItemsPerGroup = 128;   // should be a power of two for interal needs
     const uint32_t _maxLocalBuffer       = 30000; // should be less than a half of local memory (two buffers)
     const uint32_t _preferableSubGroup   = 16;    // preferable maximal sub-group size
+    const uint32_t _nPartialCentroids    = 128;
+    const uint32_t _nValuesInBlock       = 1024 * 1024 * 1024 / sizeof(algorithmFPType);
+    const uint32_t _nMinRows             = 1;
```

Comment:
It'd be good to have an explanation about these magic constants at least how it has done at lines 94-96.